apps/emqx_bridge/src/emqx_bridge_resource.erl

```diff
@@ -261,7 +267,12 @@ recreate(Type, Name, Conf, Opts) ->
 create_dry_run(Type, Conf0) ->
     TmpName = iolist_to_binary([?TEST_ID_PREFIX, emqx_utils:gen_id(8)]),
     TmpPath = emqx_utils:safe_filename(TmpName),
-    Conf = emqx_utils_maps:safe_atom_key_map(Conf0),
+    %% Some config keys might be maps that we don't want to covert to atoms.  This depends
+    %% on the schema of each bridge type.
+    Opts = #{
+        excluded_paths => config_keys_to_skip_atom_conversion(Type)
+    },
+    Conf = emqx_utils_maps:safe_atom_key_map(Conf0, Opts),
```

Comment:
Do I understand correctly that we have a raw config here converted to some kind of a "translated" config?

Is that convertion to atom keys a shortcut made instead of converting with schema 🤔 
Could we just make a full translation instead?

Reply:
It should have already been validated by the API calls.

This was already like this for quite some time.  :thinking:

Reply:
> Could we just make a full translation instead?

Sorry, what do you mean by a full translation?

Reply:
I mean, that in normal flow
* raw config comes through the API
* it is passed to `emqx_conf:update`
* callbacks receive translated (in terms of hocon) config, i.e. with atom keys, converted values, etc.
* callbacks start a new bridge

Here, as far as I understand, the dry-run checker receives a raw config, but does not send it to `emqx_conf:update` because it just wants to check resource connectivity. So it runs `emqx_utils_maps:safe_atom_key_map` as a shortcut for translation, just to make keys as atoms to make it possible to pass as the options to the resource.

My concern is that extending this shortcut usage, we start to build and support some sort of "parallel semi-transltion" mechanism which approximates `hocon:translate`. 😅 

Maybe we could do something like `Conf = hocon_tconf:check_plain(SomeSchema, RawConf, #{atom_key => true}, [Root])` instead.

Reply:
That's a much better solution, indeed.  I've updated the branch to use that approach.  Thank you!  :beers:

Reply:
Unfortunately, this seems to have consequences in some bridges that have validators and pattern matches that expect atoms or binaries in some places. :see_no_evil: 
I'll have to hunt and fix those.

Reply:
Hopefully, those should be fixed now.  They were mostly places where `map()` values were expected and some validators that matched on raw binary keys.  Added a few more tests on those.  :crossed_fingers: